File: Server/src/zone.ts
import type { Player } from "./player.js";
import { ResourceNode, spawnResourceNodes } from "./resource-node.js";
import {
  ZoneType,
  type ResourceNodeState,
  type PlayerState,
  type ZoneInfoPayload,
} from "./types.js";

export interface ZoneBounds {
  xMin: number;
  xMax: number;
  yMin: number;
  yMax: number;
}

export interface ZoneConfig {
  id: string;
  name: string;
  type: ZoneType;
  bounds: ZoneBounds;
  connections: string[]; // IDs of adjacent zones
  nodeCount: number;
}

export class Zone {
  readonly id: string;
  readonly name: string;
  readonly type: ZoneType;
  readonly bounds: ZoneBounds;
  readonly connections: string[];

  private players = new Map<number, Player>();
  private resourceNodes = new Map<number, ResourceNode>();

  constructor(config: ZoneConfig) {
    this.id = config.id;
    this.name = config.name;
    this.type = config.type;
    this.bounds = config.bounds;
    this.connections = config.connections;

    // Spawn resource nodes within bounds
    const nodes = spawnResourceNodes(config.nodeCount, this.bounds);
    for (const node of nodes) {
      this.resourceNodes.set(node.id, node);
    }
  }

  // --- Player management ---

  addPlayer(player: Player): void {
    this.players.set(player.id, player);
    player.zoneId = this.id;
  }

  removePlayer(playerId: number): Player | undefined {
    const player = this.players.get(playerId);
    if (player) {
      // Stop mining if they were mining
      player.miningNodeId = null;
      this.players.delete(playerId);
    }
    return player;
  }

  getPlayer(playerId: number): Player | undefined {
    return this.players.get(playerId);
  }

  getPlayers(): IterableIterator<Player> {
    return this.players.values();
  }

  getActivePlayers(): Player[] {
    return [...this.players.values()].filter((p) => !p.isDisconnected);
  }

  getPlayerCount(): number {
    return this.players.size;
  }

  hasPlayer(playerId: number): boolean {
    return this.players.has(playerId);
  }

  // --- Resource node management ---

  getNode(nodeId: number): ResourceNode | undefined {
    return this.resourceNodes.get(nodeId);
  }

  getNodes(): IterableIterator<ResourceNode> {
    return this.resourceNodes.values();
  }

  getNodeStates(): ResourceNodeState[] {
    return [...this.resourceNodes.values()].map((n) => n.toState());
  }

  /** Process respawn timers for all depleted nodes. Returns IDs of nodes that just respawned. */
  tickRespawns(): number[] {
    const respawned: number[] = [];
    for (const node of this.resourceNodes.values()) {
      if (node.tickRespawn()) {
        respawned.push(node.id);
      }
    }
    return respawned;
  }

  // --- State snapshots ---

  getPlayerStates(): PlayerState[] {
    return this.getActivePlayers().map((p) => p.state);
  }

  toZoneInfo(): ZoneInfoPayload {
    return {
      zoneId: this.id,
      zoneName: this.name,
      zoneType: this.type,
      bounds: { ...this.bounds },
      players: this.getPlayerStates(),
      resourceNodes: this.getNodeStates(),
    };
  }

  /** Returns a random spawn position within the zone bounds. */
  randomSpawnPosition(): [number, number] {
    const x =
      this.bounds.xMin +
      Math.random() * (this.bounds.xMax - this.bounds.xMin);
    const y =
      this.bounds.yMin +
      Math.random() * (this.bounds.yMax - this.bounds.yMin);
    return [x, y];
  }

  /** Check if a position is outside this zone's bounds. */
  isOutOfBounds(x: number, y: number): boolean {
    return (
      x < this.bounds.xMin ||
      x > this.bounds.xMax ||
      y < this.bounds.yMin ||
      y > this.bounds.yMax
    );
  }
}

File: ServerRS/src/game/zone.rs
use std::collections::HashMap;

use rand::Rng;

use crate::msg::types::{PlayerState, ResourceNodeState, ZoneBoundsPayload, ZoneInfoPayload};

use super::player::Player;
use super::resource_node::{spawn_resource_nodes, ResourceNode};

#[derive(Debug, Clone)]
pub struct ZoneBounds {
    pub x_min: f64,
    pub x_max: f64,
    pub y_min: f64,
    pub y_max: f64,
}

pub struct ZoneConfig {
    pub id: String,
    pub name: String,
    pub zone_type: String, // "space", "station", "planet"
    pub bounds: ZoneBounds,
    pub connections: Vec<String>,
    pub node_count: usize,
}

pub struct Zone {
    pub id: String,
    pub name: String,
    pub zone_type: String,
    pub bounds: ZoneBounds,
    pub connections: Vec<String>,
    players: HashMap<u32, Player>,
    resource_nodes: HashMap<u32, ResourceNode>,
}

impl Zone {
    pub fn new(config: ZoneConfig) -> Self {
        let nodes = spawn_resource_nodes(config.node_count, &config.bounds);
        let mut resource_nodes = HashMap::new();
        for node in nodes {
            resource_nodes.insert(node.id, node);
        }

        Self {
            id: config.id,
            name: config.name,
            zone_type: config.zone_type,
            bounds: config.bounds,
            connections: config.connections,
            players: HashMap::new(),
            resource_nodes,
        }
    }

    // ─── Player management ───────────────────────────────────────────

    pub fn add_player(&mut self, mut player: Player) {
        player.zone_id = self.id.clone();
        self.players.insert(player.id, player);
    }

    pub fn remove_player(&mut self, player_id: u32) -> Option<Player> {
        if let Some(mut player) = self.players.remove(&player_id) {
            player.mining_node_id = None;
            Some(player)
        } else {
            None
        }
    }

    pub fn get_player(&self, player_id: u32) -> Option<&Player> {
        self.players.get(&player_id)
    }

    pub fn get_player_mut(&mut self, player_id: u32) -> Option<&mut Player> {
        self.players.get_mut(&player_id)
    }

    pub fn get_active_players(&self) -> Vec<&Player> {
        self.players
            .values()
            .filter(|p| !p.is_disconnected)
            .collect()
    }

    pub fn get_all_players(&self) -> impl Iterator<Item = &Player> {
        self.players.values()
    }

    // ─── Resource node management ────────────────────────────────────

    pub fn get_node(&self, node_id: u32) -> Option<&ResourceNode> {
        self.resource_nodes.get(&node_id)
    }

    pub fn get_node_mut(&mut self, node_id: u32) -> Option<&mut ResourceNode> {
        self.resource_nodes.get_mut(&node_id)
    }

    #[cfg(test)]
    pub fn add_node(&mut self, node: ResourceNode) {
        self.resource_nodes.insert(node.id, node);
    }

    pub fn get_node_states(&self) -> Vec<ResourceNodeState> {
        self.resource_nodes.values().map(|n| n.to_state()).collect()
    }

    pub fn tick_respawns(&mut self) -> Vec<u32> {
        let mut respawned = Vec::new();
        for node in self.resource_nodes.values_mut() {
            if node.tick_respawn() {
                respawned.push(node.id);
            }
        }
        respawned
    }

    // ─── State snapshots ─────────────────────────────────────────────

    pub fn get_player_states(&self) -> Vec<PlayerState> {
        self.get_active_players()
            .iter()
            .map(|p| p.state.clone())
            .collect()
    }

    pub fn to_zone_info(&self) -> ZoneInfoPayload {
        ZoneInfoPayload {
            zone_id: self.id.clone(),
            zone_name: self.name.clone(),
            zone_type: self.zone_type.clone(),
            bounds: ZoneBoundsPayload {
                x_min: self.bounds.x_min,
                x_max: self.bounds.x_max,
                y_min: self.bounds.y_min,
                y_max: self.bounds.y_max,
            },
            players: self.get_player_states(),
            resource_nodes: self.get_node_states(),
        }
    }

    pub fn random_spawn_position(&self) -> (f64, f64) {
        let mut rng = rand::thread_rng();
        let x = rng.gen_range(self.bounds.x_min..self.bounds.x_max);
        let y = rng.gen_range(self.bounds.y_min..self.bounds.y_max);
        (x, y)
    }

    pub fn is_out_of_bounds(&self, x: f64, y: f64) -> bool {
        x < self.bounds.x_min || x > self.bounds.x_max || y < self.bounds.y_min || y > self.bounds.y_max
    }
}
